File: apps/api/routers/kb.py
"""封印库：树/卡/询问只读；点头与封印只写台账或路径上的假设卡。"""

from __future__ import annotations

import sys
from pathlib import Path

from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field

_PACKAGES = Path(__file__).resolve().parents[3] / "packages"
if str(_PACKAGES) not in sys.path:
    sys.path.insert(0, str(_PACKAGES))

from sealed_kb.ask import ask, suggestions
from sealed_kb.store import (
    card_view,
    domain_tree,
    load,
    ops_cards,
    ops_projections,
    ops_stream_tree,
    ops_unattached,
    runtime_tree,
)
from sealed_kb.workbench import (
    DEFAULT_THEME,
    can_nod_role,
    can_seal_role,
    journal_entries,
    list_themes,
    pending_as_dict,
    probe_as_dict,
    record_nod,
    seal_leaf,
)

router = APIRouter(prefix="/api/kb", tags=["kb"])


class AskBody(BaseModel):
    question: str = Field(min_length=1)
    role: str | None = None


class NodBody(BaseModel):
    叶子: str = Field(min_length=1)
    主题: str = DEFAULT_THEME
    结果: str = Field(min_length=1)
    断在: str = ""
    栏: str = ""
    探针: list[str] = Field(default_factory=list)


class SealBody(BaseModel):
    叶子: str = Field(min_length=1)
    主题: str = DEFAULT_THEME


def _ok(data):
    return {"code": 0, "message": "", "data": data}


def _role(x_demo_role: str | None) -> str:
    return x_demo_role or ""


@router.get("/tree")
def get_tree():
    return _ok({"roots": domain_tree(), "runtime": runtime_tree(), "readonly": True})


@router.get("/ops")
def get_ops():
    return _ok(
        {
            "roots": ops_stream_tree(),
            "unattached": ops_unattached(),
            "projections": ops_projections(),
            "cards": ops_cards(),
            "readonly": True,
        }
    )


@router.get("/card/{card_id}")
def get_card(card_id: str):
    kb = load()
    fact = kb.facts.get(card_id)
    if fact is None:
        raise HTTPException(status_code=404, detail="卡片不存在")
    return _ok(card_view(fact, kb))


@router.get("/suggestions")
def get_suggestions():
    return _ok({"questions": suggestions()})


@router.post("/ask")
def post_ask(body: AskBody):
    result = ask(body.question, body.role)
    return _ok(
        {
            "verdict": result.verdict,
            "question": result.question,
            "intent": result.intent,
            "mode": result.mode,
            "slots": result.slots,
            "missing": result.missing,
            "claim": result.claim,
            "chain": result.chain,
            "text": result.text,
            "suggestions": result.suggestions,
            "workbench_kind": result.workbench_kind,
            "workbench_theme": result.workbench_theme,
            "workbench_id": result.workbench_id,
            "readonly": True,
        }
    )


@router.get("/themes")
def get_themes():
    return _ok({"themes": list_themes(), "default": DEFAULT_THEME, "readonly": True})


@router.get("/pending")
def get_pending(theme: str = DEFAULT_THEME):
    try:
        return _ok(pending_as_dict(theme))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/probe")
def get_probe(theme: str = DEFAULT_THEME):
    try:
        return _ok(probe_as_dict(theme))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/journal")
def get_journal():
    return _ok({"entries": journal_entries(), "readonly": True})


@router.post("/nod")
def post_nod(body: NodBody, x_demo_role: str | None = Header(default=None, alias="X-Demo-Role")):
    role = _role(x_demo_role)
    if not can_nod_role(role):
        raise HTTPException(status_code=403, detail="无权点头")
    try:
        row = record_nod(
            leaf=body.叶子,
            theme=body.主题,
            result=body.结果,
            role=role,
            break_layer=body.断在,
            field=body.栏,
            probes=body.探针,
        )
    except (KeyError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _ok({"row": row, "sealed": False})


@router.post("/seal")
def post_seal(body: SealBody, x_demo_role: str | None = Header(default=None, alias="X-Demo-Role")):
    role = _role(x_demo_role)
    if not can_seal_role(role):
        raise HTTPException(status_code=403, detail="无权封印")
    try:
        changed = seal_leaf(body.叶子, body.主题, role)
    except PermissionError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    return _ok({"changed": changed})

File: web/src/api/kb.ts
import { request, requestWithRole } from "./client";

export type KbNode = {
  id: string;
  name: string;
  layer: string;
  status: string;
  pack: string;
  children: KbNode[];
  capabilities?: KbNode[];
  owners?: KbNode[];
};

export type KbCard = {
  编号: string;
  名称: string;
  套: string;
  层: string;
  状态: string;
  命题?: string | null;
  定义?: string | null;
  属于?: string | null;
  关联?: string[];
  是一种?: string[];
  别名?: string[];
  其下?: string[];
  撑住谁?: string[];
  谁有权?: string[];
  谁兑现?: string[];
  角色?: string | null;
  持有?: string | null;
  可以?: string[];
  阶段产出?: string | null;
  绑哪个对象?: string[];
  对外怎么说?: Record<string, string>;
  说明?: string | null;
  出处?: string[] | string | null;
  类型?: string | null;
  必须有?: string | boolean | null;
  禁止?: string[] | null;
};

export type AskResult = {
  verdict: string;
  question: string;
  intent: string | null;
  mode: string;
  slots: string[];
  missing: string[];
  claim: string | null;
  chain: string[];
  text: string;
  suggestions: string[];
  workbench_kind?: string;
  workbench_theme?: string;
  workbench_id?: string;
  readonly: boolean;
};

export type PendingItem = {
  shallow_id: string;
  name: string;
  status: string;
  layer: string;
  track: string;
  chain: string[];
  captured: string[];
  permits: string[];
  speech: string;
};

export type ProbeItem = {
  anchor: string;
  name: string;
  field: string;
  question: string;
  status: string;
  layer: string;
};

export type JournalEntry = {
  编号?: string;
  小白问?: string;
  状态?: string;
  锚点?: string;
  探针栏?: string;
  增量?: { 新卡?: string[]; 改栏?: { 编号?: string; 栏?: string; 到?: string[] }[]; 新边?: unknown[] };
};

export type ThemeRow = {
  主题: string;
  轨: string;
  默认打开: boolean;
  说明: string;
};

export function fetchKbTree(): Promise<{ roots: KbNode[]; runtime?: KbNode[]; readonly: boolean }> {
  return request("/api/kb/tree");
}

export function fetchKbOps(): Promise<{
  roots: KbNode[];
  unattached?: KbNode[];
  projections: KbCard[];
  cards: KbCard[];
  readonly: boolean;
}> {
  return request("/api/kb/ops");
}

export function fetchKbCard(id: string): Promise<KbCard> {
  return request(`/api/kb/card/${encodeURIComponent(id)}`);
}

export function fetchKbSuggestions(): Promise<{ questions: string[] }> {
  return request("/api/kb/suggestions");
}

export function askKb(question: string, role?: string): Promise<AskResult> {
  return request("/api/kb/ask", {
    method: "POST",
    body: JSON.stringify({ question, role }),
  });
}

export function fetchKbThemes(): Promise<{ themes: ThemeRow[]; default: string; readonly: boolean }> {
  return request("/api/kb/themes");
}

export function fetchKbPending(theme: string): Promise<{ theme: string; readonly: boolean; items: PendingItem[] }> {
  return request(`/api/kb/pending?theme=${encodeURIComponent(theme)}`);
}

export function fetchKbProbe(theme: string): Promise<{ theme: string; readonly: boolean; items: ProbeItem[] }> {
  return request(`/api/kb/probe?theme=${encodeURIComponent(theme)}`);
}

export function fetchKbJournal(): Promise<{ entries: JournalEntry[]; readonly: boolean }> {
  return request("/api/kb/journal");
}

export function nodKb(
  role: string,
  body: { 叶子: string; 主题: string; 结果: string; 断在?: string; 栏?: string; 探针?: string[] },
): Promise<{ row: Record<string, unknown>; sealed: boolean }> {
  return requestWithRole("/api/kb/nod", role, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function sealKb(role: string, body: { 叶子: string; 主题: string }): Promise<{ changed: string[] }> {
  return requestWithRole("/api/kb/seal", role, {
    method: "POST",
    body: JSON.stringify(body),
  });
}
